### PR TITLE
feat: Android App Onboarding

### PR DESCRIPTION
**Pull request by AmarTrebinjac**
## Changes

Added android app CTA to onboarding

## Events

Did you introduce any new tracking events?

## Experiment

`onboarding_android`

## Manual Testing

> [!CAUTION]
> Please make sure existing components are not breaking/affected by this PR

### Preview domain
https://mi-630.preview.app.daily.dev

### CODE REVIEW

File: packages/shared/src/components/onboarding/OnboardingAndroidApp.tsx [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    <>
+      <Image
+        className="absolute bottom-0 left-0 w-full"
+        src={cloudinaryAndroidApp320w}
+      />
+      <OnboardingTitle className="z-1">
+        Byte-sized dev news on-the-go, download our mobile app!
+      </OnboardingTitle>
+      <Button
+        tag="a"
```
Comment (rebelchris) [inline packages/shared/src/components/onboarding/OnboardingAndroidApp.tsx:19]: Shouldn't this target blank maybe?

File: packages/shared/src/components/onboarding/OnboardingHeader.tsx [hunk 2]
```diff
@@ -59,13 +63,18 @@ export const OnboardingHeader = ({
...
+              onClick={onClick}
               customActionName={customActionName}
               activeScreen={activeScreen}
             />
           )}
+          {showNotNowButton.includes(activeScreen) && (
+            <Button onClick={onClick} variant={ButtonVariant.Tertiary}>
+              {customActionName}
+            </Button>
+          )}
```
Comment (rebelchris) [inline packages/shared/src/components/onboarding/OnboardingHeader.tsx:77]: why do we need this, the above already supports customActionName?
What's different?

If it's just variant, we should make that dynamic in the above one I feel like.

File: packages/shared/src/components/profile/ProfileSettingsMenu.tsx [hunk 3]
```diff
@@ -72,6 +74,16 @@ const useMenuItems = (): NavItemProps[] => {
         }
       : undefined;
 
+    const downloadAndroidApp =
+      checkIsBrowser(UserAgent.Android) && !isAndroidApp()
```
Comment (rebelchris) [inline packages/shared/src/components/profile/ProfileSettingsMenu.tsx:78]: Is't also an experiment right?

File: packages/shared/src/components/profile/ProfileSettingsMenu.tsx [hunk 4]
```diff
@@ -72,6 +74,16 @@ const useMenuItems = (): NavItemProps[] => {
...
+        ? {
+            label: 'Download mobile app',
+            icon: <GooglePlayIcon />,
+            href: process.env.NEXT_PUBLIC_ANDROID_APP,
+            target: '_blank',
```
Comment (rebelchris) [inline packages/shared/src/components/profile/ProfileSettingsMenu.tsx:83]: Do we auto add `rel` if this set? (inside the render, else we'll need to pass it)

File: packages/shared/src/lib/func.ts [hunk 5]
```diff
@@ -38,6 +38,12 @@ export const postWindowMessage = (
 export const checkIsExtension = (): boolean => !!process.env.TARGET_BROWSER;
 export const isExtension = !!process.env.TARGET_BROWSER;
 
+export const isAndroidApp = (): boolean => globalThis?.isAndroidApp;
+
+export const isPWA = (): boolean => globalThis?.isPWA;
+
+export const isApp = isPWA() || isAndroidApp();
```
Comment (rebelchris) [inline packages/shared/src/lib/func.ts:45]: What do you use this for? I don't see any usage is that correct?

File: packages/webapp/pages/onboarding.tsx [hunk 6]
```diff
@@ -143,6 +150,10 @@ export function OnboardPage(): ReactElement {
     feature: featureOnboardingSources,
     shouldEvaluate: shouldEnrollSourceSelection,
   });
+  const appExperiment = useConditionalFeature({
+    feature: feature.onboardingAndroid,
+    shouldEvaluate: checkIsBrowser(UserAgent.Android),
```
Comment (rebelchris) [inline packages/webapp/pages/onboarding.tsx:155]: We should also add the trigger that it should only evaluate if android browser + closer to the step, basically you can add the evaluation we use for source onboarding for example.

File: packages/webapp/pages/onboarding.tsx [hunk 7]
```diff
@@ -201,12 +220,11 @@ export function OnboardPage(): ReactElement {
...
 
-    return router.replace({
-      pathname: '/',
-      query: {
-        ua: 'true',
-      },
-    });
+    if (appExperiment && activeScreen !== OnboardingStep.AndroidApp) {
+      return setActiveScreen(OnboardingStep.AndroidApp);
+    }
```
Comment (rebelchris) [inline packages/webapp/pages/onboarding.tsx:225]: It will conflict with @ilasw code, we need to be careful which order they come in.

File: packages/webapp/pages/onboarding.tsx [hunk 8]
```diff
@@ -309,14 +330,21 @@ export function OnboardPage(): ReactElement {
         showOnboardingPage={showOnboardingPage}
         setAuth={setAuth}
         customActionName={customActionName}
-        onClickCreateFeed={onClickCreateFeed}
+        onClick={
+          activeScreen === OnboardingStep.AndroidApp
+            ? onCompleteOnboarding
```
Comment (rebelchris) [inline packages/webapp/pages/onboarding.tsx:335]: We should never call this directly I think, it now misses the settings set, right?

File: packages/webapp/pages/onboarding.tsx [hunk 9]
```diff
@@ -309,14 +330,21 @@ export function OnboardPage(): ReactElement {
...
       <div
         className={classNames(
-          'flex w-full flex-grow flex-col flex-wrap justify-center px-4 tablet:flex-row tablet:gap-10 tablet:px-6',
+          'flex w-full flex-col flex-wrap justify-center px-4 tablet:flex-row tablet:gap-10 tablet:px-6',
           activeScreen === OnboardingStep.Intro && wrapperMaxWidth,
-          !isAuthenticating && 'mt-7.5 flex-1 content-center',
+          !isAuthenticating &&
+            activeScreen !== OnboardingStep.AndroidApp &&
+            'mt-7.5 flex-1 content-center',
+          activeScreen !== OnboardingStep.AndroidApp && 'flex-1 flex-grow',
```
Comment (rebelchris) [inline packages/webapp/pages/onboarding.tsx:347]: Why are these changes needed, it feels like it could affect some stuff and i don't think it should be needed, but maybe you can tell me what didn't work?

File: packages/webapp/pages/onboarding.tsx [hunk 10]
```diff
@@ -358,11 +386,16 @@ export function OnboardPage(): ReactElement {
...
             {activeScreen === OnboardingStep.Sources && <Sources />}
+            {activeScreen === OnboardingStep.AndroidApp && (
+              <OnboardingAndroidApp />
+            )}
           </div>
         )}
       </div>
       {showOnboardingPage && <OnboardingFooter />}
-      <FooterLinks className="mx-auto pb-6" />
+      {activeScreen !== OnboardingStep.AndroidApp && (
```
Comment (rebelchris) [inline packages/webapp/pages/onboarding.tsx:396]: Maybe worth checking with design if it's intentional?
Just mention every step now includes it.

**Review by rebelchris — CHANGES_REQUESTED** [review]
Some minor comments, overall looks good :)

File: packages/shared/src/components/profile/ProfileSettingsMenu.tsx (hunk 3 above)
Comment (AmarTrebinjac) [reply packages/shared/src/components/profile/ProfileSettingsMenu.tsx:78]: @rebelchris I'm actually not entirely sure if this is also part of the experiment or not, since it's in a completely different part of the app. 

I'll ask Product

File: packages/shared/src/lib/func.ts [hunk 11]
```diff
@@ -144,3 +151,13 @@ export const initReactModal = ({
 
   globalThis.reactModalInit = true;
 };
+
+export const initApp = (): void => {
+  const param = globalThis?.location?.search;
+  if (param?.includes('android=true')) {
```
Comment (capJavert) [inline packages/shared/src/lib/func.ts:157]: let's parse it as `URLSearchParams` and check with `get`, string compare can get it wrong.

File: packages/webapp/.env [hunk 12]
```diff
@@ -15,3 +15,5 @@ NEXT_PUBLIC_GB_DEV_MODE=false
 
 NEXT_PUBLIC_SLACK_CLIENT_ID=1137730955072.7361269413077
 NEXT_PUBLIC_PADDLE_TOKEN=topsecret
+
+NEXT_PUBLIC_ANDROID_APP="https://play.google.com/store/apps/details?id=dev.daily"
```
Comment (capJavert) [inline packages/webapp/.env:19]: Added this to Vercel as well?

File: packages/shared/src/lib/func.ts [hunk 13]
```diff
@@ -144,3 +151,13 @@ export const initReactModal = ({
...
+    globalThis.isAndroidApp = true;
+  }
+  if (param?.includes('pwa=true')) {
+    globalThis.isPWA = true;
+  }
```
Comment (capJavert) [inline packages/shared/src/lib/func.ts:162]: Yeah this might get cleared on full page reload, for example after oauth flow.

Comment (AmarTrebinjac) [reply packages/shared/src/lib/func.ts:162]: @capJavert An alternative would be to add it to localStorage instead I guess. What do you think?

Comment (capJavert) [reply packages/shared/src/lib/func.ts:162]: Issue is that anyone could add those params and then start to act as if they were in android app or pwa, it would be much better if we could find a way to use feature detection instead. Did you try to look into if we can detect pwa and android app through some specific browser features/APIs they have?

Comment (AmarTrebinjac) [reply packages/shared/src/lib/func.ts:162]: Yes, I did. Didn't seem to be any real common consensus on the best way to do it. For the Android app, we could alternatively set a custom user agent in the Java code before launching the webview, but seems like a lot of effort for not a lot of gain.

Maybe I'm not seeing the bigger picture, but not sure why anyone would be editing those, or what the benefit of doing it would be? The purpose of the value is purely UX

File: packages/webapp/pages/onboarding.tsx (hunk 8 above)
Comment (AmarTrebinjac) [reply packages/webapp/pages/onboarding.tsx:336]: @rebelchris We're not calling it differently than we used to, just checking which function to pass?

This mutation is still getting called if that's what you're referring to:

Screenshot 2024-11-28 at 11 31 20

Comment (rebelchris) [reply packages/webapp/pages/onboarding.tsx:336]: How though?
You either call `onCompleteOnboarding` or `onClickCreateFeed` now?
I'm not sure why it needs to be something different either way.

It's common practice already we have steps you can optionally skip it shouldn't need new logic right?

File: packages/webapp/pages/_app.tsx [hunk 14]
```diff
@@ -94,6 +95,8 @@ function InternalApp({ Component, pageProps, router }: AppProps): ReactElement {
   }, [updateCookieBanner, user]);
 
   useEffect(() => {
+    initApp();
```
Comment (sshanzel) [inline packages/webapp/pages/_app.tsx:98]: Probably better to have this in its own useEffect as this one can run more than once.

File: packages/shared/src/lib/func.ts (hunk 13 above)
Comment (capJavert) [reply packages/shared/src/lib/func.ts:162]: I think its always good to look at it from perspective of someone messing with it 😆 

Because in the future we might rely on this to do some more crucial or collect analytics data and then the data will not be accurate due to above method not being reliable. 

For android it seems like a good way to inject custom user agent since we can control it and injecting user agent is much harder then adjusting the URL in browser.

For PWA I did a google and there is this method, did you try it already?
```js
const isIos = /iPhone|iPad|iPod/.test(navigator.userAgent);
const isInStandaloneMode = window.navigator.standalone || window.matchMedia('(display-mode: standalone)').matches;

if (isIos && isInStandaloneMode) {
    console.log('PWA on iOS detected');
}
```

Seems to work if I test it with current PWA.

That said your implementation might be acceptable but I think all stakeholders and product need to be made aware of the downsides.

File: packages/webapp/pages/onboarding.tsx (hunk 7 above)
Comment (sshanzel) [reply packages/webapp/pages/onboarding.tsx:226]: We need to check with product team on how we want these onboarding UIs to be displayed.

**Review by sshanzel — COMMENTED** [review]
Don't have much to add on what's already mentioned. Looking great so far.

File: packages/shared/src/lib/func.ts (hunk 13 above)
Comment (sshanzel) [reply packages/shared/src/lib/func.ts:162]: I agree with the iOS PWA. That snippet from Ante works well. Not sure though when it comes to Android.

Comment (sshanzel) [reply packages/shared/src/lib/func.ts:162]: Most I see from the internet also checks the user agent when it comes to Android. Probably a helpful resource: https://stackoverflow.com/questions/77982241/how-can-i-detect-if-im-in-an-app-or-im-in-a-browser-on-android-with-javascript

Comment (AmarTrebinjac) [reply packages/shared/src/lib/func.ts:162]: > I think its always good to look at it from perspective of someone messing with it 😆
> 
> Because in the future we might rely on this to do some more crucial or collect analytics data and then the data will not be accurate due to above method not being reliable.
> 
> For android it seems like a good way to inject custom user agent since we can control it and injecting user agent is much harder then adjusting the URL in browser.
> 
> For PWA I did a quick google and there is this method, did you try it already?
> 
> ```js
> const isIos = /iPhone|iPad|iPod/.test(navigator.userAgent);
> const isInStandaloneMode = window.navigator.standalone || window.matchMedia('(display-mode: standalone)').matches;
> 
> if (isIos && isInStandaloneMode) {
>     console.log('PWA on iOS detected');
> }
> ```
> 
> Seems to work if I test it with current PWA.
> 
> That said your implementation might be acceptable but I think all stakeholders and product need to be made aware of the downsides.

I did see this myself, but when I was looking at this yesterday, it seemed a bit concerning that it doesn't even have a proper entry into MDN:
Screenshot 2024-11-28 at 13 57 24

Seems a tad better than the check for ?pwa=true since the browser sets it itself, but going the localStorage route seems more ideal to keep the check for PWA and App more consistent. We can decide on the PWA approach in Slack sinceI  preemptively added this due to the iOS onboarding being the next task. Not crucial for this task.

I think adjusting the user agent from the Java code is a bit out of scope for this experiment as well. Wouldn't you agree?  

@sshanzel We have a function called `isWebView` that covers that user agent. I did bring it up in guild-web, but I'm not entirely sure if it's reliable enough, since webview can be used for a variety of things. Could be a good temporary implementation for now, though.

File: packages/webapp/pages/onboarding.tsx (hunk 8 above)
Comment (AmarTrebinjac) [reply packages/webapp/pages/onboarding.tsx:336]: @rebelchris Sorry, I might've misunderstood what exactly you were referring to. I thought you were talking about how we pass the function to the component, not that you meant we shouldn't call `onCompleteOnboarding` itself explicitly.

I primarily did this to not re-call the setting of the settings (what a sentence 😅) again, but if you don't think it's a big deal I'll revert this!

File: packages/webapp/pages/onboarding.tsx (hunk 9 above)
Comment (AmarTrebinjac) [reply packages/webapp/pages/onboarding.tsx:341]: @rebelchris There's a small error on my part since flex-1 shouldn't be a part of the check. 

However, I found a solution that doesn't need to shuffle around the already existing classes.

**Review by capJavert — APPROVED** [review]
Looks good code wise, did not test extensively current/new flows.  

As for Android and query param flow, since this is behind feature flag maybe best to deploy and check on Android phone if it all works.

**Comment by AmarTrebinjac** [comment]
Did a force push above here because I pushed some ios PWA changes to the wrong branch 😅

**Review by sshanzel — APPROVED** [review]
Great work! 🚀 

Just had a question on why we would still need to set a global variable with android.

File: packages/shared/src/lib/func.ts [hunk 15]
```diff
@@ -144,3 +152,10 @@ export const initReactModal = ({
 
   globalThis.reactModalInit = true;
 };
+
+export const initApp = (): void => {
```
[the remark this answers is not in the document]
Comment (AmarTrebinjac) [reply packages/shared/src/lib/func.ts:156]: @sshanzel The android app does not set any permanent indicator that it is an app as of yet like the ios PWA does. So until we implement something more robust like setting the user agent on webview launch, this will have to do!
